### PR TITLE
feat(core): Implement `project:viewer` role

### PR DESCRIPTION
**Pull request by despairblue**
## Summary

Add a project viewer role.

This role only allows the user to view workflows, credentials and executions within a project, but not to create or modify them.

## Related tickets and issues

https://linear.app/n8n/issue/PAY-1423/add-new-project-viewer-role

## Review / Merge checklist

- [x] PR title and summary are descriptive. **Remember, the title automatically goes into the changelog. Use `(no-changelog)` otherwise.** ([conventions](https://github.com/n8n-io/n8n/blob/master/.github/pull_request_title_conventions.md))
- [x] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included.
  > A bug is not considered fixed, unless a test is added to prevent it from happening again.
  > A feature is not complete without tests.

### CODE REVIEW

File: packages/cli/src/credentials/credentials.service.ts [hunk 1]
```diff
@@ -275,7 +275,7 @@ export class CredentialsService {
 
 			if (typeof projectId === 'string' && project === null) {
 				throw new BadRequestError(
-					"You don't have the permissions to save the workflow in this project.",
+					"You don't have the permissions to save the credential in this project.",
```
Comment (despairblue) [inline packages/cli/src/credentials/credentials.service.ts:278]: unrelated fix

File: packages/cli/test/integration/credentials/credentials.api.test.ts [hunk 2]
```diff
@@ -685,7 +685,7 @@ describe('POST /credentials', () => {
 			//
 			.expect(400, {
 				code: 400,
-				message: "You don't have the permissions to save the workflow in this project.",
+				message: "You don't have the permissions to save the credential in this project.",
```
Comment (despairblue) [inline packages/cli/test/integration/credentials/credentials.api.test.ts:688]: unrelated fix

File: packages/cli/test/integration/workflows/workflows.controller.ee.test.ts
Comment (despairblue) [inline packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1]: [Make sure to ignore white space when reviewing this PR.](https://github.com/n8n-io/n8n/pull/9611/files?w=1)

I combined all the PATCH suites into one describe block.

File: packages/cli/test/integration/project.api.test.ts [hunk 3]
```diff
@@ -479,261 +479,265 @@ describe('PATCH /projects/:projectId', () => {
...
-		expect(tp1Relations.find((p) => p.userId === testUser3.id)).toBeUndefined();
-
-		// Check we haven't modified the other team project
-		expect(tp2Relations.find((p) => p.userId === testUser2.id)).not.toBeUndefined();
-		expect(tp2Relations.find((p) => p.userId === testUser1.id)).toBeUndefined();
-		expect(tp2Relations.find((p) => p.userId === testUser2.id)?.role).toBe('project:editor');
-		expect(tp2Relations.find((p) => p.userId === ownerUser.id)?.role).toBe('project:editor');
-	});
-
-	test('should not add from a project adding user with an unlicensed role', async () => {
```
Comment (despairblue) [inline packages/cli/test/integration/project.api.test.ts:607]: Yeah, I don't know what this means, but judging from the body this tests that you can't assign a role that isn't licensed. 

I changed the title and amended the implementation to test editors and viewers.

File: packages/cli/test/integration/project.api.test.ts
Comment (despairblue) [inline packages/cli/test/integration/project.api.test.ts:1]: [Make sure to ignore white space when reviewing this PR.](https://github.com/n8n-io/n8n/pull/9611/files?w=1)

I combined all the PATCH suites into one describe block.

File: packages/cli/test/integration/workflows/workflows.controller.ee.test.ts
Comment (despairblue) [inline packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1]: [Make sure to ignore white space when reviewing this PR.](https://github.com/n8n-io/n8n/pull/9611/files?w=1)

I combined all the PATCH suites into one describe block.

Comment (despairblue) [inline packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1]: [Make sure to ignore white space when reviewing this PR.](https://github.com/n8n-io/n8n/pull/9611/files?w=1)

I combined all the PATCH suites into one describe block.

Comment (despairblue) [inline packages/cli/test/integration/workflows/workflows.controller.ee.test.ts:1]: [Make sure to ignore white space when reviewing this PR.](https://github.com/n8n-io/n8n/pull/9611/files?w=1)

I combined all the PATCH suites into one describe block.

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 4]
```diff
@@ -231,6 +247,22 @@ describe('GET /credentials', () => {
...
+		const savedCredential = await saveCredential(randomCredentialPayload(), {
+			project: teamProject,
+		});
+
+		const response = await testServer
+			.authAgentFor(member)
+			.get(`/credentials/${savedCredential.id}`);
+
+		expect(response.statusCode).toBe(200);
+		expect(response.body.data).toBeDefined();
```
Comment (krynble) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:263]: I'm wondering whether we should return this or not, as the UI only shows the fields if the `credential:update` scope exists, so the fields won't be visible in the UI, and returning the data is just more potential to leakage wdyt? Maybe we can remove the `data` attribute

Comment (krynble) [reply packages/cli/test/integration/credentials/credentials.api.ee.test.ts:263]: Now I'm thinking that this might be the top-level `data` and not the actual credential information, so it should be fine. I'll approve in this case.

**Review by krynble — DISMISSED** [review]

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts (hunk 4 above)
Comment (despairblue) [reply packages/cli/test/integration/credentials/credentials.api.ee.test.ts]: I amended the assertions, it now asserts that the viewer cannot see the `data.data` attribute.

**Review by krynble — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.46.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.46.0)
